src/integrations.py

```
@@ -98,3 +120,132 @@ def auxiliary_data(self) -> AuxiliaryData:
+    def _server_key(self) -> Optional[str]:
+        return self.cert_handler.key
+
+    @property
+    def _server_cert(self) -> Optional[str]:
+        return self.cert_handler.cert
+
+    @property
+    def _ca_chain(self) -> list[str]:
+        return self.cert_handler.chain
+
+    @property
+    def cert_data(self) -> CertificateData:
+        return CertificateData(
+            ca_cert=self._ca_cert,
+            ca_chain=self._ca_chain,
+            cert=self._server_cert,
+        )
+
+    def update_certificates(self) -> None:
+        if not self.cert_handler.enabled:
+            logger.debug("The certificates integration is not ready.")
+            self._remove_certificates()
+            return
+
+        if not self.certs_ready():
+            logger.debug("The certificates data is not ready.")
+            self._remove_certificates()
+            return
+
+        self._prepare_certificates()
+        self._add_certificates()
+
+    def certs_ready(self) -> bool:
+        return all((self._ca_cert, self._ca_chain, self._server_key, self._server_cert))
+
+    def _prepare_certificates(self) -> None:
+        SERVER_CA_CERT.write_text(self._ca_cert)  # type: ignore[arg-type]
+        SERVER_KEY.write_text(self._server_key)  # type: ignore[arg-type]
+        SERVER_CERT.write_text(self._server_cert)  # type: ignore[arg-type]
+
+        try:
+            for attempt in Retrying(
+                wait=wait_fixed(3),
+                stop=stop_after_attempt(3),
+                retry=retry_if_exception_type(subprocess.CalledProcessError),
+                reraise=True,
+            ):
+                with attempt:
+                    subprocess.run(
+                        ["update-ca-certificates", "--fresh"],
+                        check=True,
+                        text=True,
+                        capture_output=True,
+                    )
+        except subprocess.CalledProcessError as e:
+            logger.error(f"{e.stderr}")
```

If the process call failed, we need to somehow notify the caller that the certificates are not ready. We can either throw an error or return a boolean. Perhaps the error is the correct approach, I think that in case of error the charm should go to a blocked state (and even stop the workload container?).